ReactAndroid/src/main/java/com/facebook/react/animated/NativeAnimatedNodesManager.java

```diff
@@ -0,0 +1,333 @@
+        " does not exists");
+    }
+    AnimatedNode childNode = mAnimatedNodes.get(childNodeTag);
+    if (childNode == null) {
+      throw new JSApplicationIllegalArgumentException("Animated node with tag " + childNodeTag +
+        " does not exists");
+    }
+    parentNode.removeChild(childNode);
+  }
+
+  public void connectAnimatedNodeToView(int animatedNodeTag, int viewTag) {
+    AnimatedNode node = mAnimatedNodes.get(animatedNodeTag);
+    if (node == null) {
+      throw new JSApplicationIllegalArgumentException("Animated node with tag " + animatedNodeTag +
+        " does not exists");
+    }
+    if (!(node instanceof PropsAnimatedNode)) {
+      throw new JSApplicationIllegalArgumentException("Animated node connected to view should be" +
+        "of type " + PropsAnimatedNode.class.getName());
+    }
+    PropsAnimatedNode propsAnimatedNode = (PropsAnimatedNode) node;
+    if (propsAnimatedNode.mConnectedViewTag != -1) {
+      throw new JSApplicationIllegalArgumentException("Animated node " + animatedNodeTag + " is " +
+        "already attached to a view");
+    }
+    propsAnimatedNode.mConnectedViewTag = viewTag;
+  }
+
+  public void disconnectAnimatedNodeFromView(int animatedNodeTag, int viewTag) {
+    AnimatedNode node = mAnimatedNodes.get(animatedNodeTag);
+    if (node == null) {
+      throw new JSApplicationIllegalArgumentException("Animated node with tag " + animatedNodeTag +
+        " does not exists");
+    }
+    if (!(node instanceof PropsAnimatedNode)) {
+      throw new JSApplicationIllegalArgumentException("Animated node connected to view should be" +
+        "of type " + PropsAnimatedNode.class.getName());
+    }
+    PropsAnimatedNode propsAnimatedNode = (PropsAnimatedNode) node;
+    if (propsAnimatedNode.mConnectedViewTag != viewTag) {
+      throw new JSApplicationIllegalArgumentException("Attempting to disconnect view that has " +
+        "not been connected with the given animated node");
+    }
+    propsAnimatedNode.mConnectedViewTag = -1;
+  }
+
+  /**
+   * Animation loop performs two BFSes over the graph of animated nodes. We use incremented
+   * {@code mAnimatedGraphBFSColor} to mark nodes as visited in each of the BFSes which saves
+   * additional loops for clearing "visited" states.
+   *
+   * First BFS starts with nodes that are in {@code mUpdatedNodes} (that is, their value have been
+   * modified from JS in the last batch of JS operations) or directly attached to an active
+   * animation (hence linked to objects from {@code mActiveAnimations}). In that step we calculate
+   * an attribute {@code mActiveIncomingNodes}. The second BFS runs in topological order over the
+   * sub-graph of *active* nodes. This is done by adding node to the BFS queue only if all its
+   * "predecessors" have already been visited.
+   */
+  public void runUpdates(UIImplementation uiImplementation, long frameTimeNanos) {
```

Comment:
Check out "what's coming up next" section in the PR description ("Unit tests for the graph traversal algorithm") :panda_face: